clients/pkg/promtail/discovery/consulagent/consul.go

```diff
@@ -0,0 +1,574 @@
+			level.Debug(d.logger).Log(
+				"msg", "removing service since consul no longer has a record of it",
+				"name", name)
+			// Call the watch cancellation function.
+			cancel()
+			delete(services, name)
+
+			// Send clearing target group.
+			select {
+			case <-ctx.Done():
+				return
+			case ch <- []*targetgroup.Group{{Source: name}}:
+			}
+		}
+	}
+
+	// Send targetgroup with no targets if nothing was discovered.
+	if len(services) == 0 {
+		select {
+		case <-ctx.Done():
+			return
+		case ch <- []*targetgroup.Group{{}}:
+		}
+	}
+}
+
+// consulService contains data belonging to the same service.
+type consulService struct {
+	name         string
+	tags         []string
+	labels       model.LabelSet
+	discovery    *Discovery
+	client       *consul.Client
+	tagSeparator string
+	logger       log.Logger
+}
+
+// Start watching a service.
+func (d *Discovery) watchService(ctx context.Context, ch chan<- []*targetgroup.Group, name string) {
+	srv := &consulService{
+		discovery: d,
+		client:    d.client,
+		name:      name,
+		tags:      d.watchedTags,
+		labels: model.LabelSet{
+			serviceLabel:    model.LabelValue(name),
+			datacenterLabel: model.LabelValue(d.clientDatacenter),
+		},
+		tagSeparator: d.tagSeparator,
+		logger:       d.logger,
+	}
+
+	go func() {
+		ticker := time.NewTicker(d.refreshInterval)
+		defer ticker.Stop()
+		agent := srv.client.Agent()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				srv.watch(ctx, ch, agent)
+				select {
+				case <-ticker.C:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+}
+
+// Get updates for a service.
+func (srv *consulService) watch(ctx context.Context, ch chan<- []*targetgroup.Group, agent *consul.Agent) {
+	level.Debug(srv.logger).Log("msg", "Watching service", "service", srv.name, "tags", strings.Join(srv.tags, ","))
+
+	t0 := time.Now()
+	aggregatedStatus, serviceChecks, err := agent.AgentHealthServiceByName(srv.name)
+	elapsed := time.Since(t0)
+	serviceRPCDuration.Observe(elapsed.Seconds())
+
+	// Check the context before in order to exit early.
+	select {
+	case <-ctx.Done():
+		return
+	default:
+		// Continue.
+	}
+
+	if err != nil {
+		level.Error(srv.logger).Log("msg", "Error refreshing service", "service", srv.name, "tags", strings.Join(srv.tags, ","), "err", err)
+		rpcFailuresCount.Inc()
+		time.Sleep(retryInterval)
+		return
+	}
+
+	self, err := agent.Self()
+	if err != nil {
+		level.Error(srv.logger).Log("msg", "failed to get agent info from agent api", "err", err)
+		return
+	}
+	var member = consul.AgentMember{}
+	memberBytes, err := json.Marshal(self["Member"])
+	if err != nil {
+		level.Error(srv.logger).Log("msg", "failed to get member information from agent", "err", err)
+		return
+	}
+	err = json.Unmarshal(memberBytes, &member)
+	if err != nil {
+		level.Error(srv.logger).Log("msg", "failed to unmarshal member information from agent", "err", err)
+		return
+	}
+
+	nodeName := self["Config"]["NodeName"].(string)
```

Comment:
Part of me would like to see this access protected against panics, but I see the method signature doesn't support an error return argument. We could use logging + failure metrics here, but the panic on such an unexpected code path may be fine for now -- it's a quicker, cruder way to surface problems.